### ISSUE TITLE
Improve code readability to invite other developers to contribute

### ISSUE THREAD

**Issue by markmarijnissen** [amark/gun#165]
Hi,

I find myself interested, yet documentation is to broad / high-level and code to unreadable to get started or contribute.

I would love
- comments in the code
- smaller files
- browser module in multiple files (bundle with webpack)
- different organization on the server side
- unit tests

Also, I'd like the API and implementation to be clearly split. For example, use a facade-implementation pattern, or even better, a plugin-based architecture to split between API for developer and actual implemenation.

Why? I see a open ticket "Reconnect logic". I know this is already battle-tested and implemented in Socket.io. So why reinvent the wheel? I would argue that hard dependencies are unsexy for some use-cases. So a plugin-architecture or a facade-impl pattern allows for maximum flexibility in this.

What are your thoughts?

**Comment by amark** [comment]
We definitely need more low level documentation and more documentation in general. Agreed.

As far as:

 - Code comments: Yes, I was going to go back through and add comments but wanted to focus on delivering stuff for people. So yes, this is coming.
 - Smaller files: only for NodeJS. I have a strong policy that GUN core should "just work" without any build steps or advanced developer tooling. Thankfully most features are built as extensions/plugins on top of GUN core and are their own file, as well as all NodeJS related functionality is implemented separately.
 - Browser modules in multiple files. Sorry, this something I cannot accommodate because it increases the complexity for beginners. Any and all modules/plugins built on top of GUN can use webpack if desired, but GUN core has to remain buildless.
 - Server side organization: I'd love to hear your thoughts on this! Unlike the browser (which I'm pretty opinionated about), I'm very open to trying out different things.
 - Unit tests are in /tests/common.js which is also single-file because I run tests in the browser and against IE6 (although I lost my PC and haven't been able to do this lately).

API and implementation clearly separated - yes. We're working on `v0.4.x` now which we're putting all our efforts into doing full end-to-end (driver) integration testing (which requires cleaner separation of API and implementation) and load-balance/pressure-stressing/battlesuite tests.

You can already plug in Socket.io into GUN by swapping out the existing websocket/JSONP driver with a Socket.io driver. Would love to see somebody contribute a Socket.io driver!

**Comment by markmarijnissen** [comment]
Cool! Great to see there is support for these ideas.

As for server-side organisation, I'll just wait for the 0.4.x release to see how API / implementation are seperated and things are going. I guess this already improved code readability a lot. Then I'll check again and see if I can make any suggestion!

**Comment by tokyo-jesus** [comment]
> I disagree with you on a lot of points, which you may feel strongly about, but I feel this topic merits some discussion. Nothing I have to say is directed at your person or at anyone else who contributes to Gun.

It appears to me that your main goal with having a buildless Gun is to make developer's lives easier. Having tools that "just work" is quite refreshing, thank you for making this a priority. However, I feel that there are better ways to reach your goals than what is currently implemented, namely providing a comfortable development experience for both users *and* contributors.

Users of JavaScript libraries have essentially two options when it comes to adding a dependency to their project. Both treat the library, essentially, as a black box:

```js
import Gun from "gun"
const db = new Gun()
```

```js
var Gun = require("gun")
const db = new Gun()
```

```html
<!doctype html>
<html>
  <!-- This source file could be copied from the release tarball or distributed via CDN -->
  <script src="gun.js"></script>
  <script>
    const db = new Gun()
  </script>
</html>
```

The point is, library users don't know or care if there was a build step as long as they don't have to invoke it themselves. Supplying a "just works" experience is as easy as configuring the right hooks in `package.json` and checking the appropriate outputs into version control.

Build tools give you the ability to create compilation pipelines that can provide a wide range of utilities. One example that I can think of off the top of my head is that is relevant to Gun is [Closure Compiler]. This tool brings hundreds of hours of Google's engineering experience and talent to the problem of making JavaScript small and fast on a wide range of devices. [JSDoc annotations] can be used to provide type information to the compiler which is used to enable static analysis and additional optimizations.

As an engineer I don't want to argue over code style, I want to solve problems. For that very reason, one of my favorite parts of Go is `gofmt`. JavaScript has a similar tool: [standard]. While I don't personally agree with some of the rules this standard adheres to, the fact that I don't have to think about coding style or formatting in conforming projects has been a huge productivity booster. The best part is that you can check for and achieve conformity automatically, there is nothing to debate; just run the command.

Another powerful use case is hot reloading and live testing. The two most repetitive parts of a web developer's workflow can be completely automated: browser refresh and invocations of `npm test`. There are a number of ways to set this up and they are all better than doing either of these activities manually.

By integrating these tools into your project you can organize and write your code in a very clean and modular manner without sacrificing anything to usability, while making huge gains in consistency, performance, and correctness.

**TL;DR:** Build tools can provide value to the project in a lot of different ways and their complexity can easily be hidden from library users. There is some setup overhead and a small learning curve on the contributor side, but, in my experience, the benefits nearly always outweigh the cost.

I don't expect anything to change, but I hope you will consider what I have to say, as a fellow software engineer.

[Closure Compiler]: https://developers.google.com/closure/compiler/
[JSDoc annotations]: https://developers.google.com/closure/compiler/docs/js-for-compiler
[standard]: http://standardjs.com/

**Comment by amark** [comment]
You have many good points. I foresee an "inverted" build system, one that doesn't ruin productivity of people like me, but also helps people like you and Jesse that like tooling.

So instead of going from "broken javascript by default that can't run in the browser" to "just works" (not what I would call "just works"), we should come up with a split utility that goes from "just works" and splits it out into all of its different modules. GUN core is highly modular, despite it all being in one file, or despite the readability. So this is certainly doable/possible.

Syntax styles is also one of those things I believe Babel would be great for. Again an "inverted" standard. Rather than there being a standard the whole community has to adopt (potentially at the loss of productivity for the self), instead there should your own personal syntax style you like (2 spaces, 4 spaces), and whenever you open up the project... Babel transpiles the source into your style.

Obviously I don't have the time to figure out how to do these things, but I'm open for somebody to do that. I however cannot risk losing my productivity (primary contributor) for what is honestly a very arbitrary styling/formatting thing. I already tried to do webpack with Jesse and it was incredibly problematic.

I'm not saying my style/format is better/superior. It isn't. Which is why you guys should have it your way also. This is why an inverted build is the perfect compromise because it could keep everybody happy and the debate a moot point.

I know LightTable already is starting to do things like this. Who's up for the challenge? I have many concepts for how the algorithm would work.

**Comment by tokyo-jesus** [comment]
I'm looking at this from two perspectives:

One, the developer who is using Gun in his or her application. To me, this is where it is important to create a "just works" experience. It should be very easy for anyone to incorporate Gun into their project or play with it in JS Fiddle. 

Two, the developer who wants to contribute to Gun. My concern here isn't having a tool that "just works", its about catering to the reader who is hacking on it. Small, logically organized source files are much easier to reason about than one file that does everything. You've already done the hard work of breaking your code into reusable modules, build tools are easy by comparison.

Like I said, I don't want to have to think about code style, I care about solving problems, not making my text look pretty. I like `standard` and `gofmt` because they do it for me.

**Comment by amark** [comment]
(1) Absolutely agreed.

(2) Is where things get more complicated. I also absolutely agree that small, logically organized modules is much easier to reason about. However this breaks down into two sub categories:

A) Is the mental overhead for a developer, in order to contribute, require apriori knowledge of the command line, build tools, and tools that do hot code reloading and watching < swimming through a single file.

B) Above, but inverse. Swimming through a single file < mental overhead than build tools, etc.

A < B, or B < A is not an objective fact. For you, "build tools are easy by comparison" indicates you are (A) camp. I wish I was that smart, but I'm not. Meaning I fall in the (B) camp, for the following reasons (WARNING: The following confessions will make you judge me as being an incompetent developer):

1. I avoid the command line because I don't have the brainspace to remember all the necessary bash commands. Without a well-cultivated command line knowledge, I have to constantly Google to do something simple. Not only does this reduce my productivity with the amount of repeat time spent, but if I don't have internet I'm skrewed. On the contrary, GUIs are explorable, I don't need to remember what to do. They are also safe, which means I can poke at things until I get the result I want without fear of `sudo rm -rf /`. To illustrate this further see the next point.

2. I've botched git way too many times because of the command line. Like massively skrewing up production systems, to the point where even my experienced developer friends spent hours trying to even get SSH to come back up. I don't know how this was even possible. So today, I use the lovely "GitHub for X" where X is your OS. It saves my butt and productivity.

3. Build tools are procedural, not functional. See Joe Armstrong (inventor of Erlang)'s awesome talk about how even he couldn't figure out Grunt. https://www.youtube.com/watch?v=lKXe3HUG2l4 I'm much less smart and it took a genius like Jesse to show me webpack, and I still couldn't get it.

4. Hot reloading is great but too much magic for me personally. If I'm trying to test a system it is extraordinarily important that only the pieces being tested are involved. If there is some other code/script that is being injected, and I don't know its behavior, guarantees, or side effects... then the debugging nightmare is not worth it. Tools like nodemon (?) for example freak out with gun (which we need to fix), where every single time gun dumped data to disk nodemon would restart. Which basically meant every few milliseconds. Angular, while not a reloader, when I was finally convinced to try it had impossibly obscure error messages that literally can't happen with regular JS (unrelated, but related to scary creepy Angular black magic: https://www.youtube.com/watch?v=Jyhz2e4grHo). Additionally, I've also heard that sourcemaps have gotten better, but I was left scarred.

4. Unfortunately, I also hate large files - agreed on this. Which is why the code is broken down into small logical modules. However the mental overhead of a large file (for me) is significantly less than any one of these above examples.

5. However, while coding it is (for me) much easier to jump through the flow of the code inside of one file while debugging than across multiple files. Or simply put, I'd have to do too much tab-switching (which causes my brain to have to task switch also) when everything is split apart beautifully. While it is nice ot have the separation, this also reduces my productivity.

To conclude, I hope you see I have some legitimate reasons behind why some people (like my stupidity) are in the (B) camp. This does not mean we are better.

This also means we should not overrule your productivity. So I again propose that the best win-win scenario for everybody is an inverted build. We can have Jesse set up build steps that automatically populate a `/src` folder with the small logical modules from the (A) camp, and for contributors that are in the (B) camp we can automatically build the `gun.js` from the `/src` folder if there is a diff in the `/src` folder on that machine (versus a diff in the `gun.js` folder). Then we can also have a build step that takes either and closure-compiles it to `gun.min.js`.

My goal is that everybody works at there productive maximum. My goal is to be tolerant of arbitrary style/tooling so everybody can be included. But I hope that also means including us stupid (B) camp folk.

**Comment by markmarijnissen** [comment]
Just my two cents:

This debate has made me judge this project as unreliable and unprofessional. I wouldn't use this library in production. 

Why? For me it is a red flag if the project is led by a single developer that does not adhere to industry standards. I am enthousiastic about the concept and vision regarding to using graph theory for a database, but I am not enthoustiastic about being subjected about every whim of the developer. I would trust a project in where the lead developer does not have a strong idiosyncratic opion, but is commited to following industry best practices. 

Then again, it might be my opinion that is strong? Who says build tools are industry best practice? Perhaps I am the one who is guilty of having an idiosyncratic opinion...?

In all other projects (i.e. angular, riot, ionic, etc) I use and trust, build tools are essential for a professional production-quality code-base. 

For library users, the library can be made accessible using premade build. 

For contributors, they need to know how to use a build system. Gulp, grunt and webpack are easy enough use and only require Node/NPM as software to run. The entire proces for installing node and running a watch or build step can be completed in under 10 minutes. If a contributor is not able to complete these steps, I seriously doubt his capacity to understand the codebase and make a quality contribution. (In my experience, when a library gains popularity, you'll get flooded with newbie questions and low quality pull requests, which create noise and need considerable amount of moderating and responding). 

As lead develoepr, you would need to setup these build steps, which might take a little longer. An expert could set this up in 30 minutes. As a newbie, it probably takes 60-120 minutes to learn and setup.

**Comment by amark** [comment]
@markmarijnissen I understand your frustration. But yes, your comments come across (to me, personally) as incredibly attacking and self-admittedly idiosyncratic. I'm particularly emotionally hurt because I spent a lot of time trying to explain and justify my perspective and I feel like you ignored and disregarded my comments - and it is not just me, it is other (quite famous) people too like Joe Armstrong (see Point 3). I'm also hurt because your views seem extraordinarily intolerant of other people's (not just mine) style. And you judge the quality of my work not on the work itself but on how I like to work. This hurts.

Right now, the upcoming release of GUN (v0.5.0) is doing 30M+ reads/sec (50X speed of Redis) and [not yet optimized] 6k writes/sec (same as Cassandra)... all in a browser in javascript on commodity laptops. See our readthesource.io interview for more info: https://www.youtube.com/watch?v=70dn1oZQFCk .

Yet you judge the library because I don't use webpack? I'm being very raw and honest with you here, but it is comments like these that sting really badly and hurt, and make me realize that Open Source has some terrible realities to it. Would your comments be any different.... if I were a developer that used Windows? Or a developer that was female?

Why is **how I code/think/debug** offensive to you? Have I offended you? I have strongly stated numerous times that people should code the way they want to (even if they are contributing to gun's ecosystem). **You are NOT subject** to every whim of mine and I have no clue how that could even be the case, so statements like these grip me with depression since I do not know what I did to you to subject you or hurt you.

Yes, I have found that some of the industry "best practices" are based off of pulpit influence, not software quality. As proof of this please watch the readthesource interview where I show that the industry standard of using anonymous callbacks is dangerous and measurably effects performance speed. Even things like people recommending use of `Array.forEach` causes measurable performance impacts.

I cannot stop you judging me or the quality of the project because I don't use Yeoman of Webpack or whatever. I'm sorry that that is the case, I have tried to reason with you why I (and others, like Joe Armstrong) live the life we do and use the tools we do. I can't stop that, you are you, but I hope I have communicated to you just how much it hurts just because I'm in the minority of developers, and not part of the mainstream majority.

**Comment by markmarijnissen** [comment]
Woah! I did not expect you to feel hurt. Seems like my message was not received as intended. 

You've heard me say I am subjected to your whims. That is not what I meant to say. What I meant to say is: _"I have experienced other open source projects where users where subject to lead developer whims. Those communities failed to gain traction because of that, and I invested effort in contributing to a dead end"_. 

While I am still scared that this could happen, I haven't experienced "being subjected to your whims". Indeed, I think you are listening and open for feedback and input, willing to put effort in defending your position.

I also want to acknowledge all the talent, skill, ingenuity and hard work you put in this library. I find your work incredible. The stats you mention for the upcoming version are asthonishing.

And now, if you have space to listen, I'd like to clarify my own intention. 

I am not feeling offerded or hurt. I simply choose not to use your library. I wanted to write down why so you would know about this. I imagined it would be horrible if developers don't use your project but don't tell you why. You'd be left in the dark wondering why the project doesn't gain traction. (also, I am dutch, so this might be a cross-cultural thingy called "dutch honesty"). 

So while I judge the code and concept itself as brilliant, I also weigh in other factors when using open source libraries. Such as:

- Is there active development? 
- Are there multiple contributors or just one? 
- Is there documentation?
- Is it mature enough to start using?
- Is there support available? (i.e. gitter, stack overflow, active issue section)
- Does it follow industry best practice?

As stated above, I've had some bad experiences of open source projects being abonded of failing to gain traction because the lead developer had a minority opinion or was not following best practices.

So how are you now, after reading this? Still feeling hurt of are you feeling more acknowledged and hurt for the hard work and quality code you've done?

PS. I agree that for libraries where performance is important, it's better to not use best practices if they hurt performance.
PS2. Obviously I am not judging you on your sex or OS, as these are irrelevant for using the library. (Whereas a built system has an impact on using the library). You care to share what "terrible reality of OS" you mean?

**Comment by hollowdoor** [comment]
@amark Don't let people get you down. Typically people just aren't very good at the art of communication. Sometimes they are actually angry, or whiny, but you can't let them get to you either. It's not open source. It's just people, people shouldn't always be taken seriously (even in a professional context), and a lot of times people just like to argue. I just might be one of those people so don't think I'm suggesting I have the moral high ground. :)

----------------------

As far as modularity goes:

For a library I think the best option right now is [rollup](http://rollupjs.org).

The cognitive overhead of command line programs (not forgetting that programming itself is overhead) is a thing. This emphasises the problem of modularity vs. some other metric that effects ease of authoring. Right now because modules require a build step the overhead is an OK argument against it.

My suggestion if you ever decide to make gun totally modular (as in files not function scopes) is to use [pre-commit](https://github.com/observing/pre-commit), or some other similar module of which there are [many](https://www.npmjs.com/search?q=precommit) to automate the build. The build scripts will also be runnable from `node run` which is simple, and safe in my opinion. NPM scripts package field is really helpful. Webpack is just too much for publishing a library really -- when you can't know what the next trend will be anyway.

I also think that modularity increases longevity of code. It has a mental effect to make a person focus on categories of usages, and makes re-usability easier. Modularity has made me a better coder. If modules are what a person uses all the time it changes there cognition. I'm not kidding. The effects of modularity (again file type, and not scope type) are more than just organization.

Plugins are another thing brought up by the OP. I have not much to say about this except it's generally a good idea.

---------------

On the command line problem:

It seems you just don't like the command line. Personally I prefer it because it is for the most part stable. 20 years from now most commands will probably still be the same. I understand how you might be hesitant to use it though. It really depends on interests. Some people like it because it adds another aspect of automation to their tool belt. Others just see it as a nuisance, or an oddball prerequisite to programming they have to put up with.

---------------

Like the OP I do recognize your skill. Maybe you do have the right ideas about how to best code this project, but don't underestimate the needs of the community. There are objective merits to specific styles as you have shown with your example of the anonymous callbacks being slow. I think (file) modularity is one of those styles.

I also don't like "best practices". There is a fine line between best practice, and what's popular. Some educated advice just so happens to also be a best practice. I think you might be politically turned off by the idea. Try not to let social politics decide things for you. Not just what others suggest, but what you think is right could be wrong. You'll get over emotional, and emotions can cloud judgement. I know it can be tough. These kinds of things take practice too.

**Comment by BartNetJS** [comment]
Hi, what is your opinion about TypeScript.
It will increase the readability and of course type checking,...
TypeScript got a lot of attraction. Even Angular 2 is written in TypeScript.
some motivation here:
https://blog.appdynamics.com/devops/the-benefits-of-migrating-from-javascript-to-typescript/
https://dzone.com/articles/stop-writing-spaghetti-code-why-you-should-use-typ

**Comment by amark** [comment]
Massive rewrite and the new `npm run unbuild` system has significantly improved this - check out the `src/` directory!

Quick reply to a couple of things:

 - Active development was all happening on the 0.5 branch for a long time, 0.5 has now been merged into master, which makes it all visible. Yay! I'm sure we'll use branches in the future though as well, just as a note.

 - Collaboration is split up across gun's ecosystem, and even within gun core it is pretty healthy with lots of people who have fixed things. Ideally, we still want to keep gun core to being as basic as possible, and all "features" are collaborated on top. This is the UNIX/NodeJS philosophy.

 - A lot of the documentation has just been bumped up, it'll show up on the homepage website soon too.

 - The "v1.0" will be mature enough, although we already have apps running in production (including a payment system), right now v0.6 is us adding all the "correctness" tests and load tests with PANIC. We've already verified a small scale 10K client/server write syncing correctly. More on this to come.

 - Support is consistently rated by people as way above par because the community is so strong and friendly. So big thanks to the community, they're making a huge difference on this and it is really inspiring to me.

 - Industry best practices? We only use science to test our system (see our work on PANIC, PTSD, and other tools), we do not follow blindly any leaders.

**Now that gun has surpassed the popularity of others** like Cassandra, Neo4j, Hadoop, Couch, Orient, Arango and others on GitHub, and managed to survive where RethinkDB, Parse, and others have shut down. We are pretty excited to say, with confidence, we'll pull through. :) Having customers and doing government stuff really helps out with that side of the equation as well.

@hollowdoor thanks for those words! We took your advice and added a pre-publish command that automates things, making sure people can now read/look at the source code in friendly modular files. Requiring those modular files are also now compatible with build tools!

@BartNetJS fortunately, GUN is compatible with TypeScript, but unfortunately rewriting it in TypeScript would take too long. :( Feel free to port it, though (although other language support would be preferred).

Given the huge improvements we made in v0.5+ and all the contributions we've gotten in general, the active community seems to be very happy... I will be closing this thread (not the discussion though) so that "issues" reflect bugs, and the chat room can be for discussion. I'm sure there will still be some people upset, but they can make a better impact by involving themselves with the community and contributing.

Thanks again for the encouragement!

**Comment by hollowdoor** [comment]
Cool. I think there's a lot of potential in Gun.js. I'm glad to see it moving in new interesting directions. I'm preoccupied with other projects so I'm not able to contribute much. Seeing that other people are supporting this project makes me feel good.

**Comment by amark** [comment]
@hollowdoor , that is fine :) just the fact that you are around and responsive is a friendly touch. Thanks!

**Comment by BartNetJS** [comment]
@amark great to hear that you checked out TypeScript! 
Nice noted by @hollowdoor 'preoccupied'
I'm in the same boat :-)
